servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherConcatWithCompletable.java

```diff
@@ -31,20 +31,163 @@
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void onSubscribe(final Cancellable cancellable) {
+            for (;;) {
+                final Cancellable c = this.cancellable;
+                if (c == TERMINAL || c == CANCEL_COMPLETABLE) {
+                    cancellable.cancel();
+                    break;
+                } else if (cancellableUpdater.compareAndSet(this, c, cancellable)) {
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            for (;;) {
+                final Cancellable c = cancellable;
+                assert c != IGNORE_CANCEL;
+                if (FirstSubscription.class.equals(c.getClass())) {
+                    if (cancellableUpdater.compareAndSet(this, c, IGNORE_CANCEL)) {
+                        next.subscribeInternal(this);
+                        break;
+                    }
+                } else if (c == TERMINAL) {
+                    // Only propagate terminal if we were cancelled after the first source terminated. Otherwise,
+                    // we may deliver items out of order and fail the TCK tests by delivering terminal after cancel.
+                    break;
+                } else if (cancellableUpdater.compareAndSet(this, c, TERMINAL)) {
+                    target.onComplete();
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void request(final long n) {
+            Cancellable currCancellable = cancellable;
+            if (FirstSubscription.class.equals(currCancellable.getClass())) {
+                ((FirstSubscription) currCancellable).request(n);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            for (;;) {
+                final Cancellable c = cancellable;
+                if (c == TERMINAL || c == CANCEL_COMPLETABLE) {
+                    break;
+                }
+                final boolean firstCancel = FirstSubscription.class.equals(c.getClass());
+                if (cancellableUpdater.compareAndSet(this, c, firstCancel ? TERMINAL : CANCEL_COMPLETABLE)) {
+                    try {
+                        c.cancel();
+                    } finally {
+                        if (firstCancel) {
+                            next.subscribeInternal(this);
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+
+        /**
+         * We need to differentiate between the first {@link Subscriber#onSubscribe(Subscription)} and the second
+         * {@link CompletableSource.Subscriber#onSubscribe(Cancellable)}. This unique type allows distinguishing
+         * from a {@link Cancellable} that is also a {@link Subscription} passed to the second method.
+         */
+        private static final class FirstSubscription implements Subscription {
```

Comment:
Thanks for clarifying. Do we need the same trick for `ConcatSubscriber`? It may have the same issue

Reply:
We could use the same trick but it isn't strictly necessary bcz we don't need to distinguish in terminal events if we have transitioned to the second subscription (bcz we don't transition until the first one completes).